### PR TITLE
Add hfs kyber nist round three

### PR DESCRIPTION
**Pull request by david415**
Okay. This is ready for review. I got rid of all the .unwrap() calls except for the two located in the "generate" method:
```
    /// Generate a new private key.
    fn generate(&mut self, _rng: &mut dyn Random) {
        // Oqs uses their own random generator
        let kemalg = kem::Kem::new(kem::Algorithm::Kyber1024).unwrap();
        let (pk, sk) = kemalg.keypair().unwrap();
        self.alg = kemalg;
        self.pubkey = pk;
        self.privkey = sk;
    }
```

Should I change the function signature or keep it as is?

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -47,8 +47,7 @@ blake2 = { version = "0.9", optional = true }
 rand = { version = "0.7", optional = true }
 sha2 = { version = "0.9", optional = true }
 x25519-dalek = { version = "1.1", optional = true }
-pqcrypto-kyber = { version = "0.6", optional = true }
-pqcrypto-traits = { version = "0.3", optional = true }
+oqs = { version = "0.3.0", optional = true }
```
Comment (BlackHoleFox) [inline Cargo.toml:50]: Please turn off the default features. Statically linking to OpenSSL is a headache on every platform, but especially broken on Windows as [noted in this issue](https://github.com/open-quantum-safe/liboqs-rust/issues/17).

Comment (david415) [reply Cargo.toml:50]: i guess you mean the default features in oqs? i'll look into it.

Comment (david415) [reply Cargo.toml]: right right. @BlackHoleFox okay i've disabled the default features in oqs.

**Comment by david415** [comment]
Note that until https://github.com/open-quantum-safe/liboqs/issues/909 and https://github.com/cloudflare/circl/issues/213 are resolved we should NOT merge this pull-request; The issue is that the liboqs Kyber is not compatible with Cloudflare's Golang Circl library's Kyber. I want them to be compatible because I'm using them both in my software. Obviously they *should* be compatible, but alas they are not.

That having been said, please do review this PR! I will make requested corrections and maintain this branch until the above compatibility issues are resolved. Thanks and cheers!

**Comment by bwesterb** [comment]
As Thom pointed out, the current Rust crate implements Round 2 Kyber, whereas Circl has been updated to Round 3 Kyber.  You'll have to wait for a release of libels or use an older version of Circl. https://github.com/katzenpost/noise/issues/9#issuecomment-778034278

**Comment by david415** [comment]
> As Thom pointed out, the current Rust crate implements Round 2 Kyber, whereas Circl has been updated to Round 3 Kyber. You'll have to wait for a release of libels or use an older version of Circl. [katzenpost/noise#9 (comment)](https://github.com/katzenpost/noise/issues/9#issuecomment-778034278)
> 
> Note that it is quite possible there will be more breaking changes to Kyber in the future.

@bwesterb I'm hoping that the maintainer of snow @mcginty will allow us to just use the main branch of the oqs rust crate for the time being until he releases a new version number.

**Comment by david415** [comment]
I've confirmed that this branch can interoperate with the Katzenpost fork of Go noise:
https://github.com/katzenpost/noise

File: Cargo.toml [hunk 2]
```diff
@@ -25,7 +25,7 @@ libsodium-resolver = ["sodiumoxide", "byteorder"]
 libsodium-accelerated = ["libsodium-resolver", "default-resolver"]
 vector-tests = []
 hfs = []
-pqclean_kyber1024 = ["pqcrypto-kyber", "pqcrypto-traits", "hfs", "default-resolver"]
+oqs_kyber1024 = ["oqs", "hfs", "default-resolver"]
```
Comment (mcginty) [inline Cargo.toml:28]: shall we just call this `kyber1024` or `nist3_kyber1024` so we don't have to rename it the next time an underlying crypto library changes?

File: src/resolvers/default.rs [hunk 3]
```diff
@@ -452,48 +451,53 @@ impl Hash for HashBLAKE2s {
...
-        kyber1024::shared_secret_bytes()
+        self.alg.length_shared_secret()
     }
 
     /// Generate a new private key.
     fn generate(&mut self, _rng: &mut dyn Random) {
-        // PQClean uses their own random generator
-        let (pk, sk) = kyber1024::keypair();
+        // Oqs uses their own random generator
+        let kemalg = kem::Kem::new(kem::Algorithm::Kyber1024).unwrap();
```
Comment (mcginty) [inline src/resolvers/default.rs:491]: Is it necessary to create a new `Kem` every time you call `generate()`?

File: src/resolvers/default.rs [hunk 4]
```diff
@@ -504,20 +508,21 @@ impl Kem for Kyber1024 {
...
     ) -> Result<(usize, usize), ()> {
-        let pubkey = kyber1024::PublicKey::from_bytes(pubkey).map_err(|_| ())?;
-        let (shared_secret, ciphertext) = kyber1024::encapsulate(&pubkey);
-        shared_secret_out.copy_from_slice(shared_secret.as_bytes());
-        ciphertext_out.copy_from_slice(ciphertext.as_bytes());
-        Ok((shared_secret.as_bytes().len(), ciphertext.as_bytes().len()))
+        let pk = self.alg.public_key_from_bytes(pubkey).ok_or_else(|| ())?;
+        self.alg.encapsulate(&pk).map_err(|_| ())?;
+        let (ciphertext, shared_secret) = self.alg.encapsulate(&pk).map_err(|_| ())?;
+        shared_secret_out.copy_from_slice(&shared_secret.into_vec());
```
Comment (mcginty) [inline src/resolvers/default.rs:514]: Looks like `Ciphertext` and `SharedSecret` implement `AsRef<[u8]>` so it seems like the `into_vec()` is unnecessary.

File: src/resolvers/default.rs [hunk 5]
```diff
@@ -504,20 +508,21 @@ impl Kem for Kyber1024 {
...
     /// Decapsulate a ciphertext producing a shared secret.
     #[must_use]
     fn decapsulate(&self, ciphertext: &[u8], shared_secret_out: &mut [u8]) -> Result<usize, ()> {
-        let ciphertext = kyber1024::Ciphertext::from_bytes(ciphertext).map_err(|_| ())?;
-        let shared_secret = kyber1024::decapsulate(&ciphertext, &self.privkey);
-        shared_secret_out.copy_from_slice(shared_secret.as_bytes());
-        Ok(shared_secret.as_bytes().len())
+        let ciphertext = self.alg.ciphertext_from_bytes(ciphertext).ok_or_else(|| ())?;
+        let shared_secret = self.alg.decapsulate(&self.privkey, &ciphertext).map_err(|_| ())?;
+        shared_secret_out.copy_from_slice(&shared_secret.into_vec());
```
Comment (mcginty) [inline src/resolvers/default.rs:524]: Same comment about `into_vec()` being unnecessary.

**Review by mcginty — CHANGES_REQUESTED** [review]

File: Cargo.toml (hunk 2 above)
Comment (david415) [reply Cargo.toml:28]: sounds like a good idea to me

File: src/resolvers/default.rs (hunk 5 above)
Comment (david415) [reply src/resolvers/default.rs:524]: agreed.

File: src/resolvers/default.rs (hunk 4 above)
Comment (david415) [reply src/resolvers/default.rs:514]: yes i think you're right.

File: src/resolvers/default.rs (hunk 3 above)
Comment (david415) [reply src/resolvers/default.rs:491]: ah good point. i think it is not necessary to create it each time generate is called.

**Comment by mcginty** [comment]
> > As Thom pointed out, the current Rust crate implements Round 2 Kyber, whereas Circl has been updated to Round 3 Kyber. You'll have to wait for a release of libels or use an older version of Circl. [katzenpost/noise#9 (comment)](https://github.com/katzenpost/noise/issues/9#issuecomment-778034278)
> > Note that it is quite possible there will be more breaking changes to Kyber in the future.
> 
> @bwesterb I'm hoping that the maintainer of snow @mcginty will allow us to just use the main branch of the oqs rust crate for the time being until a new version of the oqs crate is released. Or better yet, a specific commit id from the main branch that way it won't just break unexpectedly one day.

Given that Kyber support is still an experiment hidden behind a non-default feature flag, this seems fine.

**Comment by david415** [comment]
@mcginty Alright I've made the corrections you specified but now I'm wondering if you can help me figure out why the build fails:
```
user@computer:~/code/snow$ cargo build --features nist3_kyber1024
warning: panic message contains braces
   --> src/params/patterns.rs:580:48
    |
580 |         "handshake messages contain one of the {{'e1', 'ekem1'}} tokens, but not the other",
    |                                                ^^             ^^
    |
    = note: `#[warn(non_fmt_panic)]` on by default
    = note: this message is not used as a format string, but will be in a future Rust edition
help: add a "{}" format string to use the message literally
    |
580 |         "{}", "handshake messages contain one of the {{'e1', 'ekem1'}} tokens, but not the other",
    |         ^^^^^

warning: 1 warning emitted

    Finished dev [unoptimized + debuginfo] target(s) in 0.16s
```

**Comment by mcginty** [comment]
> @mcginty Alright I've made the corrections you specified but now I'm wondering if you can help me figure out why the build fails:
> 
> ```
> user@computer:~/code/snow$ cargo build --features nist3_kyber1024
> warning: panic message contains braces
>    --> src/params/patterns.rs:580:48
>     |
> 580 |         "handshake messages contain one of the {{'e1', 'ekem1'}} tokens, but not the other",
>     |                                                ^^             ^^
>     |
>     = note: `#[warn(non_fmt_panic)]` on by default
>     = note: this message is not used as a format string, but will be in a future Rust edition
> help: add a "{}" format string to use the message literally
>     |
> 580 |         "{}", "handshake messages contain one of the {{'e1', 'ekem1'}} tokens, but not the other",
>     |         ^^^^^
> 
> warning: 1 warning emitted
> 
>     Finished dev [unoptimized + debuginfo] target(s) in 0.16s
> ```

That's a benign warning - your build is failing on windows because of `liboqs` (I imagine it was failing before your change as well - I just enabled Windows CI a few hours ago):

https://github.com/mcginty/snow/pull/106/checks?check_run_id=1897933780#step:4:661

**Comment by BlackHoleFox** [comment]
The current issue the CI failures are seeing is due to `bindgen`. There's an issue similar to it [here](https://github.com/KyleMayes/clang-sys/issues/121). Maybe we could try pointing it at the runner's clang installation in the environment variable?

**Comment by david415** [comment]
> The current issue the CI failures are seeing is due to `bindgen`. There's an issue similar to it [here](https://github.com/KyleMayes/clang-sys/issues/121). Maybe we could try pointing it at the runner's clang installation in the environment variable?

I have no idea how to do that.

File: tests/general.rs [hunk 6]
```diff
@@ -253,10 +253,8 @@ fn test_Xpsk0_expected_value() {
 
 #[test]
 #[cfg(feature = "hfs")]
-#[cfg(feature = "pqclean_kyber1024")]
+#[cfg(feature = "nist3_kyber1024")]
 fn test_NNhfs_sanity_session() {
-    // Due to how PQClean is implemented, we cannot do deterministic testing of the protocol.
-    // Instead, we will see if the protocol runs smoothly.
```
Comment (dsprenkels) [inline tests/general.rs:259]: Why remove this comment? liboqs also incorporates its own random number generator, so this comment is still applicable.

File: Cargo.toml [hunk 7]
```diff
@@ -47,8 +47,7 @@ blake2 = { version = "0.9", optional = true }
 rand = { version = "0.8", optional = true }
 sha2 = { version = "0.9", optional = true }
 x25519-dalek = { version = "1.1", optional = true }
-pqcrypto-kyber = { version = "0.6", optional = true }
-pqcrypto-traits = { version = "0.3", optional = true }
+oqs = { git = "https://github.com/open-quantum-safe/liboqs-rust.git", rev = "6653b262d8548a8c93000a5b847b65c9b51e6e78", optional = true, default-features = false }
```
Comment (dsprenkels) [inline Cargo.toml:50]: This is dangerous, in the context of publishing on <https://crates.io>. crates.io does not allow dependencies that are pulled from a Git repo. (See also [RFC 2141](https://rust-lang.github.io/rfcs/2141-alternative-registries.html).) There has to be a crate here.

**Review by dsprenkels — CHANGES_REQUESTED** [review]

File: tests/general.rs (hunk 6 above)
Comment (david415) [reply tests/general.rs:259]: ah okay i'll put it back

File: Cargo.toml (hunk 7 above)
Comment (david415) [reply Cargo.toml:50]: Right. Can we convince the maintainers of the oqs crate to adopt a reasonable versioning scheme which progresses faster than every 6 months?

Comment (david415) [reply Cargo.toml:50]: I'd really rather not have to fork oqs and fork snow. But the way things are now it's looking like I'm going to have to in order to actually do a "cargo publish". So sad.

**Comment by mcginty** [comment]
Closing, since it's aged long enough.

**Pull request closed without merging**